runtime/common/marshaling/config.ts

```diff
@@ -0,0 +1,17 @@
+// This file is trying to create a global flag that toJSON overrides know to apply or not.
+// It's no-op for some objects but will change the toJSON format for Date.
+// The flag is not designed to be used by marshalers running in parallel. It's probably okay
+// if marshalers don't yield explicitly. In most cases JS runs with no parallelism.
+let _isMarshaling: boolean = false;
```

Comment:
Not sure I quite follow the need for the global flag and overriding prototype.toJSON. Doesn't passing a custom serializer to JSON.stringify() alone provide the necessary context and functionality to do custom marshaling?

Reply:
The JS JSON.stringify lib works in a weird way. 
1. it first checks if the targeting object has a toJSON method. If there is, the library will call the toJSON function first.
2. then it calls the replacer to parse the intermediate result. 
3. lastly it serializes the result in the typical JSON way. 

For example, since Buffer.prototype.toJSON is defined, the replacer in step 2 will never receive an object that's instance of Buffer. The object will be serialized to an object before the replacer receives it. 

--- 

A few more things. 
- The protocol of toJSON is not symmetric in general since JSON.parse has nothing like fromJSON.
- there's no context passed all the way down to toJSON() so that we can't tell if toJSON() is called by the marshaler, or the user code. that's why we need a global flag here.

Reply:
Other libraries. like jackson-js, allows passing down a stringify context, however that context is not passed to toJSON function.

Reply:
I can add these to the code doc if that makes it more clear.

Reply:
Ah got it.

Can we instead just have the marshaler recursively traverse objects and convert Errors/Dates/Buffers to your new object representation, and then just call normal JSON.stringify() once at the end?

In theory what you have is probably safe, but it seems a little scary to be replacing the native toJSON() impls any time this code gets imported.

Reply:
ok let me do that. it might be okay with a simplified version that only do the replacing for recognizable structures. at least it should work for our own use case.